File: samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/play/infrastructure/zlink/sessions/handlers/AuthenticatePlaySessionHandler.java
package systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.sessions.handlers;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import systems.zlink.framework.actors.ActorRef;
import systems.zlink.framework.actors.ZLinkActorCreateResult;
import systems.zlink.framework.actors.ZLinkActorManager;
import systems.zlink.framework.channels.ZLinkClient;
import systems.zlink.framework.streams.ZLinkSessionContext;
import systems.zlink.framework.streams.ZLinkSessionDispatchContext;
import systems.zlink.framework.streams.ZLinkTypedSessionPacketHandler;
import systems.zlink.samples.tictactoe.server.configuration.SampleNames;
import systems.zlink.samples.tictactoe.shared.contracts.AuthenticatePlayerReq;
import systems.zlink.samples.tictactoe.shared.contracts.AuthenticatePlayerRes;
import systems.zlink.samples.tictactoe.shared.contracts.AuthenticateReq;
import systems.zlink.samples.tictactoe.shared.contracts.AuthenticateRes;
import systems.zlink.samples.tictactoe.shared.contracts.PlayerActorCreateReq;

import java.util.concurrent.CompletionStage;

// --8<-- [start:doc-session-auth]
public final class AuthenticatePlaySessionHandler
        implements ZLinkTypedSessionPacketHandler<ZLinkSessionContext, AuthenticateReq> {
    private static final Logger LOGGER =
            LoggerFactory.getLogger(AuthenticatePlaySessionHandler.class);

    private final ZLinkActorManager actors;
    private final ZLinkClient channels;

    public AuthenticatePlaySessionHandler(ZLinkActorManager actors, ZLinkClient channels) {
        this.actors = actors;
        this.channels = channels;
    }

    @Override
    public Class<AuthenticateReq> messageType() {
        return AuthenticateReq.class;
    }

    @Override
    public CompletionStage<Void> handle(
            ZLinkSessionContext context,
            ZLinkSessionDispatchContext dispatch,
            AuthenticateReq request) {
        if (request.accessToken() == null || request.accessToken().isBlank()) {
            throw new IllegalArgumentException("access token is required");
        }
        return channels.requestToChannel(
                        SampleNames.ApiChannel, new AuthenticatePlayerReq(request.accessToken()))
                .timeout(SampleNames.RequestTimeout)
                .submit(AuthenticatePlayerRes.class)
                // --8<-- [start:doc-ttt-session-bind]
                .thenCompose(
                        authenticated ->
                                actors.getOrCreate(
                                                authenticated.player().actorId(),
                                                SampleNames.PlayActor)
                                        .request(new PlayerActorCreateReq(authenticated.player()))
                                        .submit()
                                        .thenCompose(
                                                result -> {
                                                    ActorRef resolvedActor = requireActor(result);
                                                    return context.actors()
                                                            .bind(requireActor(result))
                                                            .thenCompose(
                                                                    boundActor -> {
                                                                        if (!boundActor
                                                                                .ref()
                                                                                .equals(
                                                                                        resolvedActor)) {
                                                                            throw new IllegalStateException(
                                                                                    "Bound ActorRef"
                                                                                        + " does"
                                                                                        + " not match"
                                                                                        + " the resolved"
                                                                                        + " ActorRef"
                                                                                        + " for '"
                                                                                            + authenticated
                                                                                                    .player()
                                                                                                    .actorId()
                                                                                            + "'.");
                                                                        }
                                                                        if (result
                                                                                instanceof
                                                                                ZLinkActorCreateResult
                                                                                        .Existing) {
                                                                            LOGGER.info(
                                                                                    "tictactoe-lifecycle"
                                                                                        + " actor-bound"
                                                                                        + " actor={}",
                                                                                    boundActor
                                                                                            .actorId());
                                                                        }
                                                                        return context.client()
                                                                                .reply(
                                                                                        new AuthenticateRes(
                                                                                                authenticated
                                                                                                        .player()))
                                                                                .submit();
                                                                    });
                                                }));
        // --8<-- [end:doc-ttt-session-bind]
    }

    private static ActorRef requireActor(ZLinkActorCreateResult result) {
        if (result instanceof ZLinkActorCreateResult.Existing existing) {
            return existing.actor();
        }
        if (result instanceof ZLinkActorCreateResult.Created created) {
            return created.actor();
        }
        throw new IllegalStateException("Play actor creation was rejected.");
    }
}
// --8<-- [end:doc-session-auth]

File: samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/play/infrastructure/zlink/sessions/handlers/AuthenticatePlaySessionHandler.kt
package systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.sessions.handlers

import kotlinx.coroutines.future.await
import org.slf4j.LoggerFactory
import systems.zlink.framework.actors.ActorRef
import systems.zlink.framework.actors.ZLinkActorCreateResult
import systems.zlink.framework.actors.ZLinkActorManager
import systems.zlink.framework.channels.ZLinkClient
import systems.zlink.framework.kotlin.ZLinkSuspendingTypedSessionPacketHandler
import systems.zlink.framework.kotlin.kotlin
import systems.zlink.framework.streams.ZLinkSessionContext
import systems.zlink.framework.streams.ZLinkSessionDispatchContext
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleNames
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.AuthenticatePlayerReq
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.AuthenticatePlayerRes
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.AuthenticateReq
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.AuthenticateRes
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlayerActorCreateReq

// --8<-- [start:doc-session-auth]
class AuthenticatePlaySessionHandler(
    private val actors: ZLinkActorManager,
    private val channels: ZLinkClient,
) : ZLinkSuspendingTypedSessionPacketHandler<ZLinkSessionContext, AuthenticateReq> {
    override fun packetName(): String = "AuthenticateReq"

    override fun messageType(): Class<AuthenticateReq> = AuthenticateReq::class.java

    override suspend fun handle(
        context: ZLinkSessionContext,
        dispatch: ZLinkSessionDispatchContext,
        request: AuthenticateReq,
    ) {
        require(request.accessToken.isNotBlank()) { "access token is required" }
        val authenticated =
            channels
                .requestToChannel(
                    SampleNames.ApiChannel,
                    AuthenticatePlayerReq(request.accessToken),
                )
                .timeout(SampleNames.RequestTimeout)
                .submit(AuthenticatePlayerRes::class.java)
                .await()
        // --8<-- [start:doc-ttt-session-bind]
        val playActor =
            actors
                .kotlin()
                .getOrCreate(authenticated.player.actorId, SampleNames.PlayActor)
                .request(PlayerActorCreateReq(authenticated.player))
                .await()
        val resolvedActor = requireActor(playActor)
        val boundActor = context.actors().bind(requireActor(playActor)).await()
        // --8<-- [end:doc-ttt-session-bind]
        check(boundActor.ref() == resolvedActor) {
            "Bound ActorRef does not match the resolved ActorRef for '${authenticated.player.actorId}'."
        }
        if (playActor is ZLinkActorCreateResult.Existing) {
            logger.info("tictactoe-lifecycle actor-bound actor={}", boundActor.actorId())
        }
        context.client().reply(AuthenticateRes(authenticated.player)).submit().await()
    }

    private fun requireActor(result: ZLinkActorCreateResult): ActorRef =
        when (result) {
            is ZLinkActorCreateResult.Existing -> result.actor
            is ZLinkActorCreateResult.Created -> result.actor
            is ZLinkActorCreateResult.Rejected ->
                throw IllegalStateException("Play actor creation was rejected.")
        }

    private companion object {
        val logger = LoggerFactory.getLogger(AuthenticatePlaySessionHandler::class.java)
    }
}
// --8<-- [end:doc-session-auth]
